src/DeviceRendererFactory.js

```diff
@@ -210,39 +190,36 @@ module.exports = class DeviceRendererFactory {
      * @param  {DeviceRenderer}     instance The DeviceRenderer instance reference to link into each plugin.
      * @param  {Object}             options  Various configuration options.
      */
-    addPlugins(instance, options) {
+    addPlugins(instance) {
         /*
          * Load instance dedicated plugins
          */
 
         const pluginInitMap = [
-            {enabled: options.touch, class: MultiTouchEvents},
-            {enabled: options.fullscreen, class: Fullscreen},
-            {enabled: options.clipboard, class: Clipboard, params: [options.i18n]},
-            {enabled: options.fileUpload, class: FileUpload, params: [options.i18n]},
-            {enabled: options.battery, class: Battery, params: [options.i18n]},
-            {enabled: options.streamBitrate, class: StreamBitrate, params: [options.i18n]},
-            {enabled: options.gps, class: GPS, params: [options.i18n, options.gpsSpeedSupport]},
-            {enabled: options.capture, class: Screencast, params: [options.i18n]},
-            {enabled: options.identifiers, class: Identifiers, params: [options.i18n]},
-            {enabled: options.network, class: Network, params: [options.i18n]},
-            {enabled: options.phone, class: Phone, params: [options.i18n]},
-            {enabled: options.baseband, class: BasebandRIL, params: [options.i18n, options.baseband]},
-            {enabled: options.streamResolution, class: StreamResolution},
-            {enabled: options.diskIO, class: IOThrottling, params: [options.i18n]},
-            {enabled: options.gamepad, class: GamepadManager},
-            {enabled: options.biometrics, class: FingerPrint},
-            {enabled: options.buttons, class: ButtonsEvents, params: [options.i18n, options.translateHomeKey]},
-            {enabled: options.camera || options.microphone, class: MediaManager},
         ];
 
         if (typeof instance.addCustomPlugins === 'function') {
-            instance.addCustomPlugins();
+            const customPlugins = instance.addCustomPlugins();
+            if (Array.isArray(customPlugins)){
+                pluginInitMap.push(...customPlugins);
+            }
         }
+
+        const dependenciesLoaded = [];
         pluginInitMap.forEach((plugin) => {
             const args = plugin.params || [];
 
             if (plugin.enabled) {
+                // load dependencies
+                if (plugin.dependencies) {
+                    plugin.dependencies.forEach((Dep) => {
+                        if (dependenciesLoaded.indexOf(Dep.name) !== -1) {
+                            return;
+                        }
+                        new Dep(instance);
+                        dependenciesLoaded.push(Dep.name);
+                    });
+                }
                 new plugin.class(instance, ...args);
```

Comment:
I think you need to add here `dependenciesLoaded.push(plugin.class);` otherwise if you load a dependency before the plugin that requires it, then it will be loaded twice.

Reply:
This way a widget couldn't be instanciate twice. We using a class structure but we bind instantiate object to the main "instance" varibale and so the widgets are use as singleton.

If we push plugin.class into dependenciesLoaded we couldn't instanciate a widget twice. Not sure that is a problem, but could be a source of bug :/